File: prototype/ios/ios/TrilingualFamily/Llm/TrilingualCard.swift
import Foundation

struct TrilingualCard: Codable {
    let object: String
    let korean: String
    let russian: String
    let english: String
    let l1Note: String
    let pronunciationAid: String
    let ageBand: String
    let safetyFlag: String

    enum CodingKeys: String, CodingKey {
        case object, korean, russian, english
        case l1Note = "l1_note"
        case pronunciationAid = "pronunciation_aid"
        case ageBand = "age_band"
        case safetyFlag = "safety_flag"
    }

    static func parse(_ raw: String) -> TrilingualCard? {
        guard let start = raw.firstIndex(of: "{"),
              let end = raw.lastIndex(of: "}"),
              start < end else { return nil }
        let json = String(raw[start...end])
        return try? JSONDecoder().decode(TrilingualCard.self, from: Data(json.utf8))
    }
}

File: prototype/android/android/app/src/main/kotlin/com/example/trilingual/llm/TrilingualCard.kt
package com.example.trilingual.llm

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

@Serializable
data class TrilingualCard(
    val `object`: String = "",
    val korean: String = "",
    val russian: String = "",
    val english: String = "",
    val l1Note: String = "",
    val pronunciationAid: String = "",
    val ageBand: String = "",
    val safetyFlag: String = "review",
) {
    val obj: String get() = `object`

    companion object {
        private val json = Json { ignoreUnknownKeys = true; isLenient = true }

        fun parseOrNull(raw: String): TrilingualCard? {
            val trimmed = extractJson(raw) ?: return null
            return runCatching {
                // tolerate snake_case from the model
                val normalized = trimmed
                    .replace("\"l1_note\"", "\"l1Note\"")
                    .replace("\"pronunciation_aid\"", "\"pronunciationAid\"")
                    .replace("\"age_band\"", "\"ageBand\"")
                    .replace("\"safety_flag\"", "\"safetyFlag\"")
                json.decodeFromString(serializer(), normalized)
            }.getOrNull()
        }

        private fun extractJson(raw: String): String? {
            val start = raw.indexOf('{')
            val end = raw.lastIndexOf('}')
            return if (start in 0 until end) raw.substring(start, end + 1) else null
        }
    }
}
